File: src/roboBayes.cpp
#define ARMA_DONT_PRINT_ERRORS
#include <RcppArmadillo.h>
#include "Model.h"
#include "Parameters.h"
#include "regressionUpdate.h"
#include "regressionProb.h"
#include "stepUK.h"
#include "spikeUpdate.h"
#include "Unit.h"
#include "updateCPs.h"

//[[Rcpp::depends(RcppArmadillo)]]
using namespace Rcpp;
using namespace arma;

/*
  Primary analysis function

  @param datapts A matrix object {n x d}. The observations for n time points.
  @param covariates A matrix object {n x k}. The covariates for n time points.
  @param prevUnit A Unit object. The results of initialization or the previous
    time point analysis
  @param model0 A Model object. The initial hyperparameter estimates.
  @param y_store The most recent Lm data points.
  @param x_store The most recent Lm covariates.
  @param time An unsigned integer object. The last time point.
  @param Parameters The analysis settings.

  returns a List object containing
    R A numeric vector object. The current posterior distribution of 
      the run length.
    RL An integer vector object. The current run lengths that are 
       retained.
    truncRind An integer vector object. The indices of the run 
              lengths retained.
    jtR A numeric vector object. The current joint distribution of the 
        run length and data.
    pars A list object. Elements are run length specific summary 
         statistics and hyperparameters. 
    cpInds An integer matrix object {length(cpthresh) x nRuns}. Each 
           element contains the most recent change point.
    lastLs A numeric vector object. Each element contains the 
           probability that a change occurred in the previous Lsearch 
           time points, delayed by cp_delay points.
    outliers An integer vector object. Time points that have been 
             identified as outliers.}
    time An integer object. The current time.
    allcov A numeric matrix object. The run length dependent covariates
          for the retained run lengths. 
    model0 A list object. The initial hyperparameters.
    RFull A numeric matrix object. The posterior distribution for each time point
    Rm The joint distribution.
    modelCP An integer vector object. The identified change points.
    modelB A cube object. The coefficients of the identified change point models.
    modelV A cube object. The variance of the identified change point models
    y_store The most recent Lm data points.
    x_store The most recent Lm covariates.
*/

List roboBayes(const arma::mat& datapts,
               const arma::mat& covariates,
               Unit& prevUnit,
               Model& model0,
               arma::mat& y_store,
               arma::mat& x_store,
               arma::mat& allcov,
               arma::uword time,
               const Parameters& parms) {

  // define variables

  int iData;

  double logEPS = log(DBL_MIN);
  double H = 1.0 / parms.lambda;

  bool oflag;

  // strictly unsigned integers

  uword nNewDays = datapts.n_rows;
  uword nPriorDays = time;
  uword tcp, tempCP;

  uvec eliminate, i1, validpts;

  urowvec tempCP1;

  // doubles

  vec predProb, rFull(prevUnit.RFull.n_rows), tempv1, tempv2;

  rowvec growR, R0(1), temp1, temp2, tR;

  mat moreData, runLengthCov;

  // customs

  Unit newUnit;

  Model update;

  // if covariates depend on run length, extract these from the
  // covariates matrix.
  if (parms.kt > 0) {
    runLengthCov = covariates.cols(parms.k-parms.kt, parms.k-1);
  } else {
    runLengthCov = zeros(0,1);
  }

  for (int it = std::max((int) nPriorDays, 1); 
           it <= ((int) nNewDays + (int) nPriorDays - 1); ++it) {

    // initialize current unit to the values of the previous unit
    newUnit = prevUnit;

    // increment time step
    time += 1;

    // identify row of data corresponding to the current time step
    iData = it - nPriorDays;

    // gaussian related updates
    if (parms.kt > 0) {
      moreData = join_rows(allcov, runLengthCov.row(iData).t());
    } else {
      moreData = join_rows(allcov, runLengthCov);
    }

    // update hyperparameters and sufficient statistics
    // function returns with an updated newUnit.pars
    regressionUpdate(datapts.row(iData).t(), 
                     covariates.row(iData).t(), 
                     model0, 
                     newUnit.pars,
                     moreData, 
                     parms); 

    // update joint distribution

    predProb = regressionProb(datapts.row(iData).t(), 
                              prevUnit.pars,
                              newUnit.pars, 
                              parms);

    temp1 = log(prevUnit.R) + predProb.t();
    temp1.elem(find(temp1 < logEPS)).fill(logEPS);

    temp2 = exp(temp1);

    temp1 = temp2 * (1.0 - H);
    
    R0(0) = sum(temp2) * H;

    newUnit.jtR = join_rows(R0, temp1);

    // update posterior distribution
    newUnit.R = newUnit.jtR / sum(newUnit.jtR);
    newUnit.R.elem(find_nonfinite(log(newUnit.R))).fill(DBL_MIN);

    // update run lengths retained
    newUnit.RL = join_rows(zeros<urowvec>(1), prevUnit.RL);
    newUnit.RL += 1;

    if (parms.getR) {
      // if requested update full posterior distribution matrix
      rFull.zeros();
      rFull.elem(newUnit.RL-1) = newUnit.R.t();
      newUnit.RFull.col(it) = rFull;
    }

    // update Rm matrix
    if (parms.getOutliers || parms.getModels) {
      // shift columns up 1, first column is 0s
      newUnit.Rm.shed_col(parms.Lm-1);
      newUnit.Rm = join_cols(zeros<rowvec>(newUnit.Rm.n_cols), newUnit.Rm);
      newUnit.Rm = join_rows(newUnit.jtR.t(), newUnit.Rm);
     
      // shift data and covariate up 1, first row is for current time point
      y_store.shed_row(parms.Lm-1);
      y_store = join_cols(datapts.row(iData), y_store);

      x_store.shed_row(parms.Lm-1);
      x_store = join_cols(covariates.row(iData), x_store);
    }

    oflag = false;

    // update the change point list if the minimum time is exceeded
    if ((int)(time-1) >= parms.cptimemin) {

      // identify change points and calculated probability of change
      // occurring in previous Lsearch time points
      updateCPs(prevUnit, newUnit, time, parms);

      if (parms.getOutliers) {
        // if probabilities used for determining outlier is requested by
        // user perform outlier detection step

        // retrieve the last change point of the largest threshold
        tempCP1 = newUnit.cpInds.tail_rows(1);
        tempCP = tempCP1.back();
        if (tempCP != 1) { 
          oflag = true;
        }

        // if an outlier was detected, newUnit will be updated
        stepUK(tempCP, newUnit, prevUnit, model0, it, parms,
               datapts, moreData, y_store, x_store);

      }

    } else {

      // if the minimum time is not yet exceeded, pad with defaults

      newUnit.cpInds.ones(parms.cpthresh.size(), 1);
      newUnit.lastLs = join_cols(prevUnit.lastLs, zeros<vec>(1));

    }

    // collect most likely models every time there is a change point
    if (parms.getModels) {

      // most recent change point for largest threshold
      tempCP1 = newUnit.cpInds.tail_rows(1);
      tempCP = tempCP1.back();

      if (oflag && tempCP == 1) {
        i1 = unique(prevUnit.cpInds);
        tempCP = i1.back();
      }

      // if change point is not 1 and time since last change point is
      // less than Lm
      if ((tempCP != 1) && ((it-tempCP) < (uword) parms.Lm)) {

        // time since change point
        tcp = it - tempCP;

        // initially, include all possible time points in window since
        // last change point
        validpts = regspace<uvec>(tcp+1, 1);

        // remove point in window that correspond to outliers
        i1 = find(newUnit.outliers >= tempCP);

        if (i1.n_elem > 0) {
          // remove outlier
          eliminate = newUnit.outliers.elem(i1) - tempCP;
          eliminate = tcp - eliminate;

          validpts = reverse(validpts);
          validpts.shed_rows(eliminate);
          validpts = reverse(validpts);
        }

        // recover sufficient statistics and hyperparameters

        update = spikeUpdate(y_store.rows(validpts-1), 
                             x_store.rows(validpts-1),
                             tcp, model0, newUnit.pars,
                             moreData, parms);

        // subset joint probability to the window
        vec jtRtemp = prevUnit.Rm(span(tcp, prevUnit.Rm.n_rows-1), tcp).as_col();

        // calculate posterior distribution for this subset
        vec Rtemp = jtRtemp / sum(jtRtemp);

        // identify max posterior
        uword imax = index_max(Rtemp);

        // store model according to change point
        uvec modInd = find(tempCP == newUnit.modelCP);

        if (modInd.n_elem > 0) {
          newUnit.modelCP(modInd[0]) = tempCP;
          newUnit.modelB.slice(modInd[0]) = update.B.slice(imax);
          newUnit.modelV.slice(modInd[0]) = update.V.slice(imax) / 
            (update.nu(imax)-parms.d-1);
        } else {
          uvec ttempCP(1, fill::value(tempCP));
          newUnit.modelCP = join_cols(newUnit.modelCP, ttempCP);

          cube tb(parms.k, parms.d, 1);
          tb.slice(0) = update.B.slice(imax);
          newUnit.modelB = join_slices(newUnit.modelB, tb);

          cube tv(parms.d, parms.d, 1);
          tv.slice(0) = update.V.slice(imax) / 
                         (update.nu(imax)-parms.d-1);
          newUnit.modelV = join_slices(newUnit.modelV, tv);
        }
      }
    }

    // truncate R and pars if necessary
    // check if the R vectors need to be truncated based on threshold
    if ((int) newUnit.R.n_elem > parms.truncRmin) {

      // remove the first truncRmin runs
      tR = newUnit.R;
      tR.shed_cols(regspace<uvec>(0, parms.truncRmin-1));

      // identify any elements of the "extra" that are at or below the threshold
      uvec id1 = find(tR <= parms.truncRthresh);

      if (id1.n_elem > 0) {
         // initialize the retained indices to be all elements of R
        newUnit.truncRind = regspace<uvec>(1, newUnit.R.n_elem);

        // remove the elements that correspond to the elements
        // not satisfying the threshold
        newUnit.truncRind.shed_rows(id1 + parms.truncRmin);

        if (parms.getOutliers || parms.getModels) {

          // normalize each column of the probabilities
          mat s1 = newUnit.Rm;
          s1.each_row() /= sum(s1);

          // initialize temp indices to be all cases and
          // remove those for which the total probability
          // of the last Lm steps is below the threshold
          uvec truncRind_m = regspace<uvec>(1, newUnit.R.n_elem);

          for (int i = s1.n_rows-1; i >= 0; --i) {
            if (all(s1.row(i) <= parms.truncRthresh)) {
              truncRind_m.shed_row(i);
            }
          }

          // incorporate any kept indices into the retained index vector
          newUnit.truncRind = sort(unique(join_cols(newUnit.truncRind, truncRind_m)));
        }

        // truncate all results to include only those identified above
        // DO NOT truncate RFull

        newUnit.R = newUnit.R.cols(newUnit.truncRind-1);

        newUnit.jtR = newUnit.jtR.cols(newUnit.truncRind-1);
        newUnit.RL = newUnit.RL.cols(newUnit.truncRind-1);

        if (parms.getOutliers || parms.getModels) {
          newUnit.Rm = newUnit.Rm.rows(newUnit.truncRind-1);
        }

        newUnit.pars.subset(newUnit.truncRind-1);

        moreData = reverse(moreData,1);
        uvec ttrunc = newUnit.truncRind;
        ttrunc.shed_row(0);
        moreData = moreData.cols(ttrunc - 2);
        moreData = reverse(moreData, 1);

      } else {
        newUnit.truncRind = regspace<uvec>(1, newUnit.R.n_elem);
      }
    } else {
      newUnit.truncRind = regspace<uvec>(1, newUnit.R.n_elem);
    }

    prevUnit = newUnit;

    allcov = moreData;
  }


  List outPars = prevUnit.pars.toList();
  List outPars0 = model0.toList();

  List toReturn = List::create(Named("RFull", prevUnit.RFull), 
                               Named("R", prevUnit.R), 
                               Named("RL", prevUnit.RL), 
                               Named("truncRind", prevUnit.truncRind),
                               Named("Rm", prevUnit.Rm),
                               Named("jtR", prevUnit.jtR),
                               Named("pars", outPars), 
                               Named("cpInds", prevUnit.cpInds),
                               Named("lastLs", prevUnit.lastLs),
                               Named("outliers", prevUnit.outliers),
                               Named("modelCP", prevUnit.modelCP),
                               Named("modelB", prevUnit.modelB),
                               Named("modelV", prevUnit.modelV),
                               Named("y_store", y_store),
                               Named("x_store", x_store),
                               Named("time", time), 
                               Named("allcov", moreData),
                               Named("model0", outPars0));

  return toReturn;
}

File: R/roboBayes.R
#' roboBayes: Robust Online Bayesian Monitoring
#'
#' The primary function for Robust Online Bayesian Monitoring (roboBayes). 
#'   Performs roboBayes on the response data regressed on the input covariates, 
#'   which should account for nonstationarity in the response between 
#'   changepoints. The algorithm outputs changepoints based on a threshold, a 
#'   roboBayes object incorporating data thus far, and algorithm settings. A 
#'   roboBayes object can be updated with new data by including it in a successive 
#'   roboBayes() function call.
#'
#' All inputs after the ellipsis are analysis settings that can only be set
#'   in the initial step of the roboBayes procedure. In the first step, input
#'   roboBayes must be NULL. All subsequent steps of the procedure must provide,
#'   through input roboBayes, the value object returned by the previous step.
#'
#' Analysis parameters must obey the following conditions:
#'   \itemize{
#'     \item Lsearch + cp_delay + Lgroup <= truncRmin + 1
#'     \item Lwindow + cp_delay + Lgroup <= truncRmin + 1
#'     \item Lgroup <= Lwindow + 1
#'     \item cptimemin > Lgroup
#'     \item cptimemin >= cp_delay - 1
#'     \item cp_delay < Lm
#'     \item Lsearch <= truncRmin
#'   }
#'
#' @param datapts A matrix object (n x d). The observations for n time points.
#'   There are methods in place to attempt to convert non-matrix input 
#'   (data.frame, numeric) to a matrix. If a numeric vector is provided, 
#'   it is assumed that d = 1.
#'
#' @param covariates A matrix object (n x k). The covariates for n time points.
#'   If any covariates depend on the run length (kt > 0), these covariates
#'   must be provided in the last kt columns of the input matrix.
#'   There are methods in place to attempt to convert non-matrix input 
#'   (data.frame, numeric) to a matrix. If a numeric vector is provided, 
#'   it is assumed that k = 1.
#'
#' @param RoboBayes A value object returned by a prior roboBayes() analysis or NULL.
#'   If NULL, the analysis is taken to be the first step, and all analysis
#'   settings (inputs after the ellipsis) can be adjusted. For all subsequent 
#'   steps, RoboBayes is the value object of the preceding step and analysis 
#'   settings are taken from that object -- any analysis setting inputs provided
#'   for continuation steps are ignored.
#'
#' @param ... Ignored. Included only to require named inputs. If RoboBayes is an
#'   object of class RoboBayes, all inputs after the ellipsis are ignored.
#' 
#' @param lambda A numeric object. The prior probability of a changepoint
#'   occurring at any time point.
#'
#' @param par_inits A list object. The initial estimates for the hyperparameters.
#'   The list can contain one or more of the following:
#'   \itemize{
#'   \item B: A (k x d) matrix. The location parameter of the matrix Normal.
#'            The default is a zero matrix.
#'
#'   \item V: A (d x d) matrix. The scale matrix of the Inverse Wishart.
#'            The default is a symmetric matrix with 1.0 on the diagonal and
#'            0.1 in the off-diagonal elements.
#'
#'   \item nu: A scalar. The degrees of freedom of the Inverse Wishart.
#'             The default is (d - 0.9).
#'
#'   \item Lambda: A (k x k) matrix. The scale parameter of the matrix Normal.
#'                 The default is a diagonal matrix of 0.01.}
#'
#' @param truncRthresh A scalar object. A probability threshold used to 
#'   limit the size of the data used at each time step. Once t > truncRmin,
#'   run lengths associated with data points from times less than t-truncRmin 
#'   for which the probability < truncRthresh are removed 
#'
#' @param truncRmin An integer object. The minimum number of time points 
#'   maintained before data truncation methods are utilized.
#'
#' @param cpthresh A numeric vector object. One or more changepoints.
#'
#' @param cptimemin An integer object. Minimum time at which updated change
#'   points can be ascertained.
#'
#' @param Lgroup An integer object. The number of runs grouped together to
#'   identify a new changepoint.
#'
#' @param Lgroup An integer object. The size of the moving window used to 
#'   aggregate run length probabilities during the initial coarse search for 
#'   the presence of a changepoint.
#'
#' @param Lsearch An integer object. The number of run length probabilities to 
#'   aggregate for calculating the probability of "recent" change.
#'
#' @param Lwindow An integer object. The number of points in the window to scan
#'   for the most recent changepoint. Changes further back than Lwindow + 
#'   cp_delay will not be recorded.
#'
#' @param Lm An integer object. The number of data points to store in memory
#'   for outlier elimination. A change recorded more than Lm points previously
#'   will not be examined for being an outlier. When getModels is TRUE, 
#'   Lm is automatically set to  Lwindow + cp_delay + 1 to ensure that model
#'   parameters associated with detected changepoints can be recovered.
#'
#' @param alpha A scalar object. The probability threshold for declaring a 
#'   point an outlier.
#'
#' @param kt An integer object. The number of covariates that depend on run
#'   length.
#'
#' @param pc A scalar object. The prior probability of the model without an
#'   outlier being the correct model, given that a candidate change/outlier has
#'   been detected.
#'
#' @param cp_delay An integer object. The minimum number of points that must
#'   be observed after the change to declare a changepoint.
#'
#' @param outlier_mean A numeric vector (d). The mean value for the 
#'   distribution of outlier values.
#'
#' @param outlier_var A matrix object (d x d). The covariance matrix for the 
#'   distribution of outlier values.
#'
#' @param getR A logical object. If TRUE, the matrix of probabilities is
#'   returned. Warning -- selecting TRUE will increase computation time, and
#'   for large analyses, may exceed available memory.
#'
#' @param getOutliers A logical object. Detect outliers. If TRUE, the matrix of
#'   probabilities used for outlier elimination are returned. Selecting TRUE
#'   will increase computation time.
#'
#' @param getModels A logical object. If TRUE, the model parameters of each
#'   changepoint are returned, as well as the matrix of probabilities used
#'   for outlier elimination. When TRUE, parameter Lm is set to 
#'   Lwindow + cp_delay + 1, regardless of that provided as input.
#'
#' @returns A RoboBayes object, which extends list and contains
#'   \itemize{
#'   \item R: A numeric vector object. The current posterior distribution of 
#'            the run length.
#'   \item RL: An integer vector object. The current run lengths that are 
#'             retained.
#'   \item truncRind: An integer vector object. The indices of the run 
#'                    lengths retained.
#'   \item jtR: A numeric vector object. The current joint distribution of the 
#'              run length and data.
#'   \item pars: A list object. Elements are run length specific summary 
#'               statistics and hyperparameters. With the exception of nu, 
#'               each hyperparameter/summary statistic is a 3 dimensional array, 
#'               where the final dimension corresponds to the run length. The 
#'               degrees of freedom, nu, are returned as a vector, of length 
#'               equal to the number of run lengths retained.
#'   \item cpInds: An integer matrix object (length(cpthresh) x nRuns). Each 
#'                 element contains the most recent changepoint.
#'   \item lastLs: A numeric vector object. Each element contains the 
#'                 probability that a change occurred in the previous Lsearch 
#'                 time points, delayed by cp_delay points.
#'   \item time: An integer object. The current time.
#'   \item allcov: A numeric matrix object. The run length dependent covariates
#'                 for the retained run lengths.
#'   \item model0: A list object. The initial hyperparameters.
#'   \item lastDataPt: A numeric vector object. The data of the last time point.
#'   \item call: The matched call.
#'   \item params: A list object. The analysis settings.
#'   }
#' Conditionally returned elements include:
#' 
#' If getR = TRUE
#'   \itemize{
#'   \item RFull: A numeric matrix object. The ith column contains the run 
#'                length posterior distribution for time i.
#'   }
#' If getOutliers = TRUE
#'   \itemize{
#'     \item Rm: A numeric matrix object. The ith column contains the joint 
#'              distribution of data and run length associated with time-i+1.
#'     \item y_store: A numeric matrix object. Contains the most recent Lm data 
#'                    points.
#'     \item x_store: A numeric matrix object. Contains the most recent Lm 
#'                    covariates.
#'     \item outliers: An integer vector object. Time points that have been 
#'                     identified as outliers.
#'   }
#' If getModels = TRUE
#'   \itemize{
#'     \item Rm: A numeric matrix object. The ith column contains the joint 
#'              distribution of data and run length associated with time-i+1.
#'     \item y_store: A numeric matrix object. Contains the most recent Lm data 
#'                    points.
#'     \item x_store: A numeric matrix object. Contains the most recent Lm 
#'                    covariates.
#'     \item mods: A list object. Each element of the list corresponds to an 
#'                 identified changepoint. For each changepoint, the expected 
#'                 model coefficients and covariance. 
#'     \item currentModel: A list object. For the current most likely run length, 
#'                    the expected model coefficients and covariance.
#'   }
#'
#'
#' @examples
#' \dontshow{
#'   RcppArmadillo::armadillo_throttle_cores(2)
#' }
#'
#'
#' nt <- 100
#'
#' ## 2 covariates each time step
#'
#' x <- cbind(rep(1,nt), rnorm(nt))
#'
#' ## 2 observations each time step
#'
#' # covariance matrix
#' sigma <- matrix(data = -0.3, nrow = 2, ncol = 2)
#' diag(sigma) <- 1.0
#'
#' # mean
#' mean1 <- c(0.5, 0.1)
#' mean2 <- c(0.2, 0.2)
#'
#' y <- rbind(mvtnorm::rmvnorm(n = nt/2, mean = mean1, sigma = 0.001*sigma),
#'            mvtnorm::rmvnorm(n = nt/2, mean = mean2, sigma = 0.001*sigma))
#'
#' # add in an outlier
#' y[30,] <- c(0.9,0.01)
#' y[y < 0.0] <- 0.0
#'
#' step1 <- roboBayes(datapts = y[1:32,], covariates = x[1:32,])
#'
#' step2 <- roboBayes(datapts = y[33:64,], covariates = x[33:64,], RoboBayes = step1)
#'
#' step3 <- roboBayes(datapts = y[65:94,], covariates = x[65:94,], RoboBayes = step2)
#'
#' step4 <- roboBayes(datapts = y[95:100,], covariates = x[95:100,], RoboBayes = step3)
#'
#' @export
#' @useDynLib roboBayes
#' @import Rcpp
#' @importFrom RcppArmadillo armadillo_throttle_cores
#' @include checkTypes.R verifyAnalysisSettings.R initRoboBayes.R regressionInit.R
#' @include RcppExports.R

roboBayes <- function(datapts,
                      covariates,
                      RoboBayes = NULL, 
                      ...,
                      lambda = 1000.0,
                      par_inits = NULL,
                      truncRthresh = 1e-4,
                      truncRmin = 100L,
                      cpthresh = 0.8,
                      cptimemin = 4L,
                      Lgroup = 3L,
                      Lsearch = 10L,
                      Lwindow = 30L,
                      Lm = 7L,
                      alpha = 0.9,
                      kt = 0L,
                      pc = 0.5,
                      cp_delay = 3L,
                      outlier_mean = rep(x = 0.5, times = ncol(x = datapts)),
                      outlier_var = diag(x = 2.0, nrow = ncol(x = datapts)),
                      getR = FALSE,
                      getOutliers = TRUE,
                      getModels = FALSE) {

  matchedCall <- match.call()

  # ensure that user did not misspell an input or provide
  # unexpected inputs
  if (length(list(...)) != 0L) {
    stop("unrecognized input provided ",
         paste(names(list(...)), collapse = ", "))
  }

  # datapts must always be provided as input
  if (is.null(x = datapts)) stop("datapts must be provided")

  # ensure that datapts is a matrix. If provided a data.frame, it is
  # converted to a data.matrix; if provided a vector, it is converted
  # to a one column matrix
  datapts <- .checkMatrix(x = datapts, name = "datapts")

  if (nrow(x = datapts) == 0L) {
    return( RoboBayes )
  }
  
  # covariates must always be provided as input
  if (is.null(x = covariates)) stop("covariates must be provided")

  # ensure that covariates is a matrix. If provided a data.frame, it is
  # converted to a data.matrix; if provided a vector, it is converted
  # to a one column matrix
  covariates <- .checkMatrix(x = covariates, name = "covariates")

  if (is.null(x = RoboBayes)) {
    # RoboBayes is null for the first step of the procedure. 

    # Ensure appropriate inputs for analysis settings.
    params <- .verifyAnalysisSettings(d = ncol(x = datapts),
                                      k = ncol(x = covariates),
                                      lambda = lambda,
                                      par_inits = par_inits,
                                      truncRthresh = truncRthresh,
                                      truncRmin = truncRmin,
                                      cpthresh = cpthresh,
                                      cptimemin = cptimemin,
                                      Lgroup = Lgroup,
                                      Lsearch = Lsearch,
                                      Lwindow = Lwindow,
                                      Lm = Lm,
                                      alpha = alpha,
                                      kt = kt,
                                      pc = pc,
                                      cp_delay = cp_delay,
                                      outlier_mean = outlier_mean,
                                      outlier_var = outlier_var,
                                      getR = getR,
                                      getOutliers = getOutliers,
                                      getModels = getModels)

    # verify or initialize initial estimates for hyper parameters
    if (is.list(x = par_inits)) {
      RoboBayes <- initRoboBayes(par_inits = par_inits, 
                                 params = params,
                                 nTimes = nrow(x = datapts))
    }

  } else if (methods::is(object = RoboBayes, class2 = "RoboBayes")) {
    # RoboBayes object indicates a continuation step. Take analysis settings
    # from previous step object
    params <- RoboBayes$params

    # ensure that data has proper number of observations for each time step
    if (ncol(x = datapts) != params$d) {
      stop("datapts is of inappropriate dimension")
    }

    # ensure that covariates has proper number of observations for each time step
    if (ncol(x = covariates) != params$k) {
      stop("covariates is of inappropriate dimension")
    }

    # remove parameters from previous object
    RoboBayes$params <- NULL

    if (params$getR) {
      # if user requested to return matrix of posterior distributions for
      # each time, extend previous result to include space for new time points

      numpts <- RoboBayes$time
  
      # number of new data points
      numNewPts <- nrow(x = datapts)

      RoboBayes$RFull <- cbind(RoboBayes$RFull, 
                               matrix(data = 0.0, 
                                      nrow = nrow(x = RoboBayes$RFull), 
                                      ncol = numNewPts))
      RoboBayes$RFull <- rbind(RoboBayes$RFull,
                               matrix(data = 0.0,
                                      nrow = numNewPts,
                                      ncol = ncol(x = RoboBayes$RFull)))
    }

  } else {
    # RoboBayes is neither NULL nor a RoboBayes object, throw error
    stop("RoboBayes is not of a recognized class")
  }

  if (is.null(x = RoboBayes)) {

    # if RoboBayes is still NULL, par_inits was not provided. Set this
    # up in the C++ code
    res <- roboBayesFirst(datapts = datapts,
                          covariates = covariates,
                          paramsList = params)

  } else if (methods::is(object = RoboBayes, class2 = "RoboBayes")) {

    # if RoboBayes is a RoboBayes object, either par_inits was provided in first
    # step or this is a continuation step

    RoboBayes <- unclass(x = RoboBayes)

    nt <- length(x = RoboBayes$mods)
    modelCP <- numeric(length = nt)
    modelB <- array(data = 0.0, dim = c(params$k, params$d, nt))
    modelV <- array(data = 0.0, dim = c(params$d, params$d, nt))
    if (nt > 0) {
      for (i in 1L:nt) {
        modelCP[i] <- RoboBayes$mods[[ i ]]$cp
        modelB[,,i] <- RoboBayes$mods[[ i ]]$B
        modelV[,,i] <- RoboBayes$mods[[ i ]]$V
      }
    }

    RoboBayes$mods <- list("cp" = modelCP, "B" = modelB, "V" = modelV)

    res <- roboBayesPrior(datapts = datapts,
                          covariates = covariates,
                          RoboBayes = RoboBayes,
                          paramsList = params)
  } else {
    # if neither condition is true, logic isn't working as expected.
    stop("unexpected RoboBayes object; contact developer")
  }

  if (params$getModels) {
    # extract the model for the most likely run length
    imax <- which.max(x = res[["R"]])
    deno <- res$pars$nu[imax] - params$d - 1.0
    res$currentModel <- list("B" = res$pars$B[,,imax],
                             "V" = res$pars$V[,,imax]/deno)

    # redefine the models for CP to conform to original definition
    if (length(res$modelCP) > 0L) {
      res$mods <- list()

      for (i in 1L:length(res$modelCP)) {
        res$mods[[ i ]] <- list("cp" = res$modelCP[i],
                                "B" = matrix(data = res$modelB[,,i],
                                             nrow = params$k,
                                             ncol = params$d),
                                "V" = matrix(data = res$modelV[,,i],
                                             nrow = params$d,
                                             ncol = params$d))
      }
    }
  }

  if (!params$getR) {
    # if posterior distribution for all time points not requested,
    # remove place holder from returned object
    res$RFull <- NULL
  }

  if (!params$getOutliers && !params$getModels) {
    # if outliers or models not requested, remove place holders from 
    # returned object
    res$Rm <- NULL
    res$y_store <- NULL
    res$x_store <- NULL
  }

  if (!params$getOutliers) {
    # if outliers not requested, remove from returned object
    res$outliers <- NULL
  }

  # remove changepoint model information from returned object
  res$modelCP <- NULL
  res$modelB <- NULL
  res$modelV <- NULL

  res$lastDataPt <- datapts[nrow(x = datapts),]
  res$call <- matchedCall
  res$params <- params
 
  class(x = res) <- c("RoboBayes", class(x = res))

  return( res )
}

